### ISSUE TITLE
Can't set format in default_url_options

### ISSUE THREAD

**Issue by nashbridges** [rails/rails#16358]
We use the old way to deliver mobile-specific content via `:format`, that is, we have `template.html.slim` and `template.mobile.slim`. We also want to persist format in URL, so a mobile user will see an `/items/10.mobile` URL instead of `/items/10`.

In Rails 3.X we could achieve this by overriding `default_url_options`, however, this is no longer work (see test example below).

I've tracked the problem down to https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/routing/route_set.rb#L260. If I extract `:format` from `@segment_keys`, then `url_for` won't receive a hash, which would otherwise include `format: nil`.

```ruby
#gem 'rails', '3.2.18' # OK
gem 'rails', '4.0.8' # FAIL
#gem 'rails', '4.1.4' # FAIL

# FAIL
#unless File.exist?('Gemfile')
  #File.write('Gemfile', <<-GEMFILE)
    #source 'https://rubygems.org'
    #gem 'rails', github: 'rails/rails'
    #gem 'arel', github: 'rails/arel'
    #gem 'rack', github: 'rack/rack'
    #gem 'i18n', github: 'svenfuchs/i18n'
  #GEMFILE

  #system 'bundle'
#end

#require 'bundler'
#Bundler.setup(:default)

require 'rails'
require 'action_controller/railtie'

class TestApp < Rails::Application
  config.root = File.dirname(__FILE__)
  config.session_store :cookie_store, key: 'cookie_store_key'
  config.secret_token    = 'secret_token' * 30
  config.secret_key_base = 'secret_key_base'

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    get '/' => 'test#index'
    resources :cars
  end
end

Car = Struct.new(:to_param)

class TestController < ActionController::Base
  include Rails.application.routes.url_helpers

  def default_url_options(options = {})
    options[:format] = params[:format] if params[:format].present?
    options
  end

  def index
    render inline: <<-VIEW
      <%= link_to "hello", car_path(Car.new(3)) %>
    VIEW
  end
end

require 'minitest/autorun'
require 'rack/test'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def test_persists_format
    get '/', format: "mobile"
    assert_equal '<a href="/cars/3.mobile">hello</a>', last_response.body.strip
  end

  private
    def app
      Rails.application
    end
end

```

**Comment by pixeltrix** [comment]
This maybe related to #18627

**Comment by nashbridges** [comment]
@pixeltrix, I confirm that mentioned PR fixes this issue

**Comment by pixeltrix** [comment]
Fixed by db870f2
